### PR TITLE
Ensure serializer throws JsonException when deserializing incompatible JSON into collections

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/corefx/issues/41839 for 5.0. https://github.com/dotnet/corefx/pull/42071 fixes it for 3.1.

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs [hunk 1]
```diff
@@ -52,6 +52,8 @@ private static void HandleStartObject(JsonSerializerOptions options, ref ReadSta
                     state.Current.ReturnValue = value;
                     state.Current.DetermineIfDictionaryCanBePopulated(state.Current.ReturnValue);
                 }
+
+                state.Current.CollectionPropertyInitialized = true;
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs]: Was this a separate issue?

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs [hunk 2]
```diff
@@ -70,6 +72,12 @@ private static void HandleEndObject(ref ReadStack state)
                 state.Current.JsonClassInfo.UpdateSortedPropertyCache(ref state.Current);
             }
 
+            if (state.Current.JsonClassInfo.ClassType == ClassType.Value)
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs]: Any reason this is beneath the `UpdateSortedPropertyCache` call above? It seems like this error check should be first.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs (hunk 1 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs]: No known/reported issues. Nothing else comes to mind beyond the current issue.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs (hunk 2 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs]: Agreed.

**Pull request merged**